### DISCUSSION TITLE
String clusters vs codepoints

### DISCUSSION THREAD

**Discussion by Andrew15-5** [typst/typst#2769] [Q&A] [answered]
Can someone explain what is the practical (Typst typesetting) difference between these two methods? It looks like they produce identical arrays (tested on some composite Unicode characters).

> ### clusters
> 
> Returns the grapheme clusters of the string as an array of substrings.
> `self.clusters()` -> [array](https://typst.app/docs/reference/foundations/array/)
> 
> ### codepoints
> 
> Returns the Unicode codepoints of the string as an array of substrings.
> `self.codepoints()` -> [array](https://typst.app/docs/reference/foundations/array/)

**Comment by emilyyyylime** [comment, marked as answer]
You say you tried 'composite' unicode characters, but you might not understand correctly what a unicode cluster is. A cluster is any one or more codepoints that combine together into a single logical character, for example the bulky "🫱🏻‍🫲🏿", consisting of five whole codepoints (four for hands and their respective skin tone modifiers, and one Zero-Width Joiner between them). You may also consider "â"; consisting of "a" and a Combining Circumflex Accent (— not to be confused with â, a single codepoint and a character in its own right)

Thus, `"🫱🏻‍🫲🏿".codepoints()` will return a five-element list, while `"🫱🏻‍🫲🏿".clusters()` will return just one.

In terms of actual typesetting value, such low-level string handling utilities don't serve any obvious purpose, probably more useful for niche library stuff

**Reply by Andrew15-5** [reply]
Hmm, I think I wasn't paying as much attention as I more or less quickly tested different stuff, like just separate chars and composite ones together. Now I clearly see the difference and the names do make sense. Thank you.
